native/dorgql.go

```diff
@@ -0,0 +1,105 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/lapack"
+)
+
+// Dorgql generates the m×n matrix Q with orthonormal columns defined as the
+// last n columns of a product of k elementary reflectors of order m as returned
+// by Dgelqf. That is,
+//  Q = H[k-1] * ... * H[1] * H[0].
+// Please see Dgelqf for more information.
+//
+// tau must have length at least k, and Dorgql will panic otherwise.
+//
+// Work is temporary storage, and lwork specifies the usable memory length. At minimum,
+// lwork >= n, and Dorgql will panic otherwise. The amount of blocking is
+// limited by the usable length.
+// If lwork == -1, instead of computing Dorgql the optimal work length is stored
+// into work[0].
+func (impl Implementation) Dorgql(m, n, k int, a []float64, lda int, tau, work []float64, lwork int) {
+	checkMatrix(m, n, a, lda)
+	if len(tau) < k {
+		panic(badTau)
+	}
+	nb := impl.Ilaenv(1, "DORGQL", " ", m, n, k, -1)
+	lworkopt := n * nb
+	work[0] = float64(lworkopt)
+	if lwork == -1 {
+		return
+	}
+	if lwork < n {
+		panic(badWork)
+	}
+	if n == 0 {
+		return
+	}
+
+	nbmin := 2
+	var nx, ldwork int
+	iws := n
+	if nb > 1 && nb < k {
+		// Determine when to cross over from blocked to unblocked code.
+		nx = max(0, impl.Ilaenv(3, "DORGQL", " ", m, n, k, -1))
+		if nx < k {
+			// Determine if workspace is large enough for blocked code.
+			ldwork = nb
```

Comment:
Work is input as a linear array, but it's used as a matrix. Fortran is column-based so the stride is n, while we're row based so the stride is nb.